src/Microsoft.Identity.Client/Telemetry.cs

```diff
@@ -48,5 +50,129 @@ public static Telemetry GetInstance()
         {
             return Singleton;
         }
+
+        public bool TelemetryOnFailureOnly { get; set; }
+
+        private Dictionary<Tuple<string, string>, EventBase> EventsInProgress = new Dictionary<Tuple<string, string>, EventBase>();
+
+        private Dictionary<string, List<EventBase>> CompletedEvents = new Dictionary<string, List<EventBase>>();
+
+        internal string GenerateNewRequestId()
+        {
+            return Guid.NewGuid().ToString();
+        }
+
+        internal void StartEvent(string requestId, EventBase eventToStart)
+        {
+            EventsInProgress[new Tuple<string, string>(requestId, eventToStart["event_name"])] = eventToStart;
+        }
+
+        internal void StopEvent(string requestId, EventBase eventToStop)
+        {
+            Tuple<string, string> eventKey = new Tuple<string, string>(requestId, eventToStop["event_name"]);
+
+            // Locate the same name event in the EventsInProgress map
+            EventBase eventStarted = null;
+            if (EventsInProgress.ContainsKey(eventKey))
+            {
+                eventStarted = EventsInProgress[eventKey];
+            }
+
+            // If we did not get anything back from the dictionary, most likely its a bug that StopEvent
+            // was called without a corresponding StartEvent
+            if (null == eventStarted)
+            {
+                // Stop Event called without a corresponding start_event.
+                return;
+            }
+
+            // Set execution time properties on the event
+            eventToStop.Stop();
+
+            if (!CompletedEvents.ContainsKey(requestId))
+            {
+                // if this is the first event associated to this
+                // RequestId we need to initialize a new List to hold
+                // all of sibling events
+                List<EventBase> events = new List<EventBase>();
+                events.Add(eventToStop);
+                CompletedEvents[requestId] = events;
+            }
+            else
+            {
+                // if this event shares a RequestId with other events
+                // just add it to the List
+                CompletedEvents[requestId].Add(eventToStop);
+            }
+
+            // Mark this event as no longer in progress
+            EventsInProgress.Remove(eventKey);
+        }
+
+        internal void Flush(string requestId)
+        {
+            if (_receiver == null)
+            {
+                return;
+            }
+
+            // check for orphaned events...
+            List<EventBase> orphanedEvents = CollateOrphanedEvents(requestId);
+            // Add the OrphanedEvents to the completed EventList
+            if (!CompletedEvents.ContainsKey(requestId))
+            {
+                // No completed Events returned for RequestId
+                return;
+            }
+
+            CompletedEvents[requestId].AddRange(orphanedEvents);
```

Comment:
Not quite understand why "all events are orphaned" would need to have special treatment, but I definitely won't mind add the new comment. :->